generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -164,6 +166,175 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from bitwise shifting.
+   */
+  private Optional<Expr> opaqueZeroOrOneBitwiseShift(BasicType type, boolean constContext,
+                                                     final int depth, Fuzzer fuzzer,
+                                                     boolean isZero) {
+    if (!BasicType.allIntegerTypes().contains(type)) {
+      return Optional.empty();
+    }
+    if (!shadingLanguageVersion.supportedBitwiseOperations()) {
+      return Optional.empty();
+    }
+    final BinOp operator = generator.nextBoolean() ? BinOp.SHL : BinOp.SHR;
+    // The maximum safe value to shift zero without potentially causing undefined behavior is
+    // between zero and eight bits (as nine bits is the minimum precision of an integer in GLSL).
+    // If we're shifting one, we want to be able to shift backwards without losing information, and
+    // so our maximum shift value is seven bits.
+    final int bound = isZero ? 10 : 9;
+    final int maxValue = generator.nextInt(bound);
+    final Expr shiftValue = new FunctionCallExpr(
+        "clamp",
+        fuzzer.fuzzExpr(type, false, constContext, depth),
+        makeOpaqueZero(type, constContext, depth, fuzzer),
+        applyIdentityFunction(
+            new TypeConstructorExpr(
+                type.toString(),
+                type.getElementType() == BasicType.INT
+                    ? new IntConstantExpr(String.valueOf(maxValue))
+                    : new UIntConstantExpr(maxValue + "u")),
+            type, constContext, depth, fuzzer));
+    if (isZero) {
+      return Optional.of(
+          new ParenExpr(
+              new BinaryExpr(
+                  makeOpaqueZero(type, constContext, depth, fuzzer),
+                  applyIdentityFunction(shiftValue, type, constContext, depth, fuzzer),
+                  operator)));
+    } else {
+      assert !isZero;
+      assert operator == BinOp.SHL || operator == BinOp.SHR;
+      // We're going to shift twice in opposite directions by the same value.
+      final BinOp secondOp = operator == BinOp.SHL ? BinOp.SHR : BinOp.SHL;
+      return Optional.of(
+          new ParenExpr(
+              new BinaryExpr(
+                  new ParenExpr(
+                      new BinaryExpr(
+                          makeOpaqueOne(type, constContext, depth, fuzzer),
+                          applyIdentityFunction(shiftValue.clone(), type, constContext, depth,
+                              fuzzer),
+                          operator)),
+                  applyIdentityFunction(shiftValue.clone(), type, constContext, depth, fuzzer),
+                  secondOp)));
+    }
+  }
+
```

Comment:
^ A general thing on the above: I'm a little worried about getting expression size explosion due to the amount of places the identity function is applied.  Perhaps don't apply the identity function to "shiftValue.clone()".

A separate point: it seems nice to have the option to shift left and then right using the same expression as shift value.  But it would also be interesting to shift left with one expression that is clamped to n, and right with a different expression that is clamped to n.  How about you decide with 0.5 probability whether to generate a new shift value expression for shifting back, vs. cloning the existing one?